### PR TITLE
variable arity

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/functions/corefuncs.py [hunk 1]
```diff
@@ -13,43 +13,51 @@
...
+
+        It is originally designed for lists of bits. It just counts the number of 1,
+        and returns len(x) - number of ones..
+        It also works in the continuous case but in that cases discretizes the
+        input domain by ]0.5,1.5] --> 1 and 0 everywhere else.
+        """
+        result = 0
+        for i, x_ in enumerate(list(x)):
+            result += 1 if (x_ == round(i % arity)) else 0
+        return len(x) - result
```
Comment (jrapin) [inline nevergrad/functions/corefuncs.py:29]: can you reimplement using arrays, they are faster and more concise

File: nevergrad/functions/corefuncs.py [hunk 2]
```diff
@@ -13,43 +13,51 @@
...
+
+        Returns len(x) - number of initial 1. I.e.
+        leadingones([0 1 1 1]) = 4,
+        leadingones([1 1 1 1]) = 0,
+        leadingones([1 0 0 0]) = 3.
+        """
+        for i, x_ in enumerate(list(x)):
+            if int(round(x_)) != i % arity:
+                return len(x) - i
+        return 0
```
Comment (jrapin) [inline nevergrad/functions/corefuncs.py:43]: use arrays

File: nevergrad/functions/corefuncs.py [hunk 3]
```diff
@@ -13,43 +13,51 @@
...
-    """
-    n = len(x)
-    m = n // 4
-    o = n - onemax(x)
-    if o == n or o <= n - m:
-        return n - m - o
-    return o  # Deceptive part.
+def discrete_functions(arity: int=2) -> tp.Any:
+    
+    def onemax(x: tp.List[int]) -> float:
```
Comment (jrapin) [inline nevergrad/functions/corefuncs.py:18]: avoid functions in functions, I suppose this should be a class:
`DiscreteFunction(arity: int, name: str)`
and should that replace leadingones5 and so on?

**Review by jrapin — CHANGES_REQUESTED** [review]

File: nevergrad/functions/corefuncs.py (hunk 3 above)
Comment (teytaud) [reply nevergrad/functions/corefuncs.py:18]: I think leadingones5 is not used.

File: nevergrad/functions/corefuncs.py (hunk 1 above)
Comment (teytaud) [reply nevergrad/functions/corefuncs.py]: The original implementation does not use arrays -- it's true that arrays would be better but this is unrelated to the present PR.

File: nevergrad/functions/corefuncs.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/functions/corefuncs.py]: The original implementation does not use arrays -- it's true that arrays would be better but this is unrelated to the present PR.

File: nevergrad/functions/corefuncs.py [hunk 4]
```diff
@@ -13,43 +13,59 @@
...
+            return 0
+
+        if self.name == "jump":
+            n = len(x)
+            m = n // 4
+            o = n - onemax(x)
+            if o == n or o <= n - m:
+                return n - m - o
+            return o  # Deceptive part.
+        assert False, f"Unknown function type {self.name}."
```
Comment (jrapin) [inline nevergrad/functions/corefuncs.py:62]:
```suggestion
       raise ValueError(f"Unknown function type {self.name}.")
```
it should probably be tested in the init though.

File: nevergrad/functions/test_corefuncs.py [hunk 5]
```diff
@@ -54,8 +54,8 @@ def test_genzcornerpeak_inf() -> None:
     genzgaussianpeakintegral=(corefuncs.genzgaussianpeakintegral, 0.10427, None),
     minusgenzgaussianpeakintegral=(corefuncs.minusgenzgaussianpeakintegral, -0.10427, None),
     linear=(corefuncs.linear, 0.57969, None),
-    onemax5=(corefuncs.onemax5, 8, np.arange(50)),
-    jump5=(corefuncs.jump5, 6, np.arange(50)),
+    onemax5=(corefuncs.onemax5, 9, np.arange(50)),
+    jump5=(corefuncs.jump5, 7, np.arange(50)),
```
Comment (jrapin) [inline nevergrad/functions/test_corefuncs.py:58]: why have they changed? That means the computation has changed

File: nevergrad/functions/corefuncs.py (hunk 3 above)
Comment (jrapin) [reply nevergrad/functions/corefuncs.py]: they should be deleted then

**Comment by teytaud** [comment]
Closed for now.

**Pull request closed without merging**